Incorporate an OTel-wide maturity matrix

bhs opened the pull request:
Per the header comment for the file under review here:

>   OpenTelemetry is a complex project with many moving parts. Thankfully, we strive to keep those parts loosely coupled where we can (this has been a goal since the very announcement of OpenTelemetry, per this blog post: https://medium.com/opentracing/merging-opentracing-and-opencensus-f0fe9c7ca6f0)
> 
>   This file describes the structure of the project and its sub-components, definitions of API and stability maturity levels, and finally records the current maturity of each component. We maintain this file as YAML in order to make it easier to automate scripts (e.g., for the website) that present the data to OpenTelemetry end-users.

This PR stems from the most recent technical *and* governance committee meetings. cc @mtwo who is doing a survey about blockers to a beta release in 5 languages for Kubecon EMEA.

All comments welcome, but especially those about the actual schema (formatting, repo location, naming, etc will be easy to change and I'm very flexible on those fronts!).

lizthegrey commented:
Fixed some typoes, updated golang status, added "unimplemented" as a state in addition to "alpha".

we currently do _not_ release major versions on breaking changes while we're in alpha for golang, as a FYI. our goal is to release major/minor versions corresponding to spec major/minor versions rather than incrementing in unpredictable ways that slew from the spec major/minor versions.

bhs commented on maturity-matrix.yaml:
```diff
@@ -47,9 +47,11 @@ productionMaturityLevels:
 
-# All OpenTelemetry APIs follow semver conventions (i.e.,
-# backwards-incompatibly changes should bump the major version number).
+# All OpenTelemetry APIs follow semver conventions (i.e., after v1.x,
```
yes, thank you for this! that's actually what I'd meant to write, so it's wonderful to see your telepathy skills on display... :)

bhs commented on maturity-matrix.yaml:
```diff
@@ -47,9 +47,11 @@ productionMaturityLevels:
+# All OpenTelemetry APIs follow semver conventions (i.e., after v1.x,
+# backwards-incompatible changes should bump the major version number).
 apiMaturityLevels:
+  # "unimplemented" APIs do not exist yet for the component/language.
```
@lizthegrey WDYT about (also) adding something like "unplanned" or "undesired" or "notApplicable"? E.g., auto-instrumentation in some future `otel-c99` support.

lizthegrey replied:
agree we should introduce n/a.

bhs replied:
done

tedsuo commented:
CC @austinlparker 

We’ve been trying to track this via github milestones, and automatically scrape that data to update the status page on the website. Would be great if we could continue to have an automated process for keeping the website up to date.

bhs commented:
@tedsuo @austinlparker yes, agree – I like the idea of using repo metadata (milestones, tags, whatever) to track status. And also agree about the website. My goal with this PR is primarily to debate "the schema" for what we're telling end-users about maturity. If we figure that out, I am personally flexible about the mechanics as long as they allow for automation and don't require some crazy amount of upfront work.

austinlparker commented:
Once we've got the schema sorted out then it should be pretty easy to add some tags or whatever to auto-populate the website.

lizthegrey commented:
Agree, the main change here is moving off of one-dimensional v0.1->v0.4 versioning to tagging each maturity stage per category and project.

SergeyKanzhelev commented on maturity-matrix.yaml:
```diff
@@ -0,0 +1,161 @@
+  - traceSDK
+  - metricsSDK
+  - contextSDK
+  - automaticInstrumentation
```
should the integrations and exporters be also tracked? Zipking and Jaeger exporters, integrations like SpringBoot, etc.?

austinlparker replied:
IMO we should focus solely on SDK/API for trace/metric/context here and allow exporters and integrations to simply reference their stability or maturity against the lower level components.

SergeyKanzhelev replied:
some integrations and exporters are the baseline for the stability as you cannot use API/SDK without ability to export it. Similar for integrations. I'm not sure how automaticInstrumentation here is different than integrations?

austinlparker replied:
Hm, that's a good point, although I don't know how to square it. Automatic instrumentation does seem to require integrations.

mtwo replied:
+1, I think that end-users expect at least some of these (OT exporter, Jaeger exporter, Prometheus exporter, basic HTTP integrations) to be tracked

bhs replied:
I do _not_ want to suggest that we try to maintain a full compatibility matrix with framework-level things like Spring. The "registry" seems like the right place to track that sort of thing.

By "automaticInstrumentation", I meant specifically the idea of "no-code" (literally) installation of the OpenTelemetry instrumentation. This is an important requirement for many brownfield services that are no longer actively maintained but nevertheless participate in transactions. As such, I would not like to extend the definition to "1-line code changes" that nevertheless require recompilation.

@SergeyKanzhelev re your comment:

>some integrations and exporters are the baseline for the stability 

I disagree – they may be the baseline for "usefulness", but not for stability. 

Given the possibly large diversity of exporters in the future, I would personally prefer that we:

1. Adopt the API and production maturity levels for exporters, and
2. Track exporter maturity via the registry

Remember that we're not debating how this information is displayed on a website at the moment, we're just debating what should be considered "internal" to OpenTelemetry rather than a plugin/etc. The website can absolutely bring in the maturity levels for popular exporters in designated languages.

SergeyKanzhelev replied:
We need to avoid situation when OT is only usable with the specific vendor exporter. This is why I think we should declare some baseline exporters here. It may be minimum functionality and not scalable, but must be reliable and stable.

austinlparker replied:
I wonder about the 'diversity of exporters' -- Yuri could probably weigh in here, but I suspect that many "vendors", both OSS and commercial, would like to simply accept OTel data format natively and thus not require an exporter at all.

With that in mind, this feels like a transient issue; Yes, there'll be maturity levels for exporters but over time those exporters simply won't matter since your backend will accept the data with no translation.

SergeyKanzhelev replied:
> I suspect that many "vendors", both OSS and commercial, would like to simply accept OTel data format natively and thus not require an exporter at all.

Not immediately. It may take a while

> With that in mind, this feels like a transient issue; Yes, there'll be maturity levels for exporters but over time those exporters simply won't matter since your backend will accept the data with no translation.

Yes, but if we intent to track progress towards 1.0 I'd suggest to include tracking of this transient issue

bhs replied:
@SergeyKanzhelev I'm going to continue to push back on this... if I want to use Jaeger's exporter and it's the only thing that's mature for my given language, IMO that's fine. Now, whether the SIGs would feel comfortable marking their APIs as "stable" (or even "beta") with zero or one exporter implemented is a different question.

SergeyKanzhelev commented on maturity-matrix.yaml:
```diff
@@ -0,0 +1,161 @@
+# end-user programming language.
+standaloneComponents:
+  collector:
+    api: beta
```
there is always a "latest stable" and "current" versions. Should we make this spearation?

mtwo replied:
In the long-term, definitely. However I don't think that we need to until after the first release.

bhs replied:
I agree with @mtwo here... for now, I think we're basically implicitly tracking "latest stable" which probably won't be very stable for a few more months. :)

tigrannajaryan replied:
Please change Collector `api` maturity to "alpha". We expect minor breaking changes to public interfaces in the Collector codebase in the near future. Once that is done we will declare it "beta".

For `production` it matches the suggested definition of "beta" (run on production at a public company) so we can keep it.

bhs replied:
@tigrannajaryan done 👍

SergeyKanzhelev commented on maturity-matrix.yaml:
```diff
@@ -0,0 +1,161 @@
+# backwards-incompatible changes should bump the major version number).
+apiMaturityLevels:
+  # "unimplemented" APIs do not exist yet for the component/language.
+  - unimplemented
```
should there also be "N/A". Do we expect auto-instrumentation for the GoLang for instance?

lizthegrey replied:
Yes, auto-instrumentation for golang includes wrappers for gorillamux, sqlx, etc. -- so it's not _fully_ automatic but it's O(1) lines of code to add.

SergeyKanzhelev replied:
Yeah. But than it falls under "integrations" bucket and we may want more than overall status? Perhaps those integrations can also contribute to the registry that @austinlparker is working on?

austinlparker replied:
Yeah, I think auto-instrumentation is just a type of integration in the registry.

bhs replied:
@lizthegrey re

>auto-instrumentation for golang includes wrappers for gorillamux, sqlx, etc. -- so it's not fully automatic but it's O(1) lines of code to add.

I'm sympathetic to the fact that it's hard to do better than this ^^ in Go, but I (unfortunately) don't think 1-line changes "count" for brownfield auto-instrumentation in many environments. This would support some sort of "n/a" designation per @SergeyKanzhelev's comment.

An alternative would be to define different levels of auto-instrumentation: perhaps "unsupported", "smallCode", and "noCode". Or, alternatively, to define auto-instrumentation criteria differently for different languages (with their different expectations).

SergeyKanzhelev replied:
Either n/a or just move it to the registry

SergeyKanzhelev commented on maturity-matrix.yaml:
```diff
@@ -0,0 +1,161 @@
+      api: unimplemented
+      production: unknown
+    automaticInstrumentation:
+      api: unimplemented
```
if it's not planned - should it be simply ommitted?

bhs replied:
I like the idea of being explicit here, or at least allowing for the option.

SergeyKanzhelev commented on maturity-matrix.yaml:
```diff
@@ -0,0 +1,161 @@
+  golang:
+    traceAPI:
+      api: beta
+      production: beta
```
I'd suggest to include the link to the issues query on GitHub showing what's left for this area. It may be an optional attribute

bhs replied:
I'm a little hesitant to do this since in the early stages, that list will have lots of errors of omission and could lead to a false sense of security...

SergeyKanzhelev replied:
can we declare it as optional so I can use it for .NET?

austinlparker replied:
IMO it'd be good to have a URL field in the schema that points to the repo.

austinlparker commented on maturity-matrix.yaml:
```diff
@@ -0,0 +1,161 @@
+  # scale, but either due to the uncertainty introduced by active development
+  # or a small number of production deployments, they should still be used
+  # cautiously in new deployments. They do not have known security issues.
+  - beta
```
I don't know if it's possible, but it might be nice to try and lock down what 'production workloads at scale' means exactly.

bhs replied:
lol, true. This was lazy on my part.

How about `... production workloads at scale: that is, the component is used in support of well-known product functionality at publicly-held companies.`

I don't love this, but am hoping that it's difficult to become publicly-held without achieving some version of "scale"... admittedly not a perfect match for the kind of scale we're seeking here, but I am loath to add a Spans/sec threshold here.

SergeyKanzhelev replied:
good enough I think

austinlparker replied:
sgtm

bhs commented:
Hi folks... I've updated the PR here in two ways:
* I removed the auto-instrumentation stuff until it's more fully-baked (it can go in the brand-new registry in the meantime)
* I added a blurb about "beta" APIs requiring multiple impls, with at least one being a major OSS project

Can reviewers either approve or request further changes? I can then fill in the gaps in other languages, potentially as follow-on PRs. cc @lizthegrey @SergeyKanzhelev @austinlparker @mtwo @tedsuo (who've all provided comments earlier – thanks).

bhs commented:
One more ping on this – still looking for an approval from an approver!

bhs commented:
@lizthegrey thanks.

All: I'll merge this tomorrow or Tuesday if there are no further concerns.

As mentioned on the most recent Gov Committee call, I'll also make a very crude sketch of how we could represent this on the marketing site and submit an issue to track that. (I don't think I'll have the cycles to actually implement the marketing site piece, though)

bhs commented:
Thanks, all... Happy to take PRs on this file as we bring it into sync with reality.
